spring-core-6/src/test/java/com/baeldung/httpinterface/BooksServiceMockitoTest.java

```diff
@@ -0,0 +1,102 @@
+package com.baeldung.httpinterface;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatusCode;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Mono;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.*;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@ExtendWith(MockitoExtension.class)
+class BooksServiceMockitoTest {
+
+    @Mock
+    private WebClient webClient;
+
+    @Mock
+    private WebClient.RequestBodySpec requestBody;
+
+    @Mock
+    private WebClient.RequestBodyUriSpec requestBodyUri;
+
+    @Mock
+    private WebClient.ResponseSpec response;
+
+    @Mock
+    private Mono<ResponseEntity<Void>> monoResponseEntity;
+
+    @Mock
+    private ResponseEntity<Void> responseEntity;
+
+    @InjectMocks
+    private BooksClient booksClient;
+
+    @Test
+    void givenMockedWebClient_whenAllBooksAreRequested_thenTwoBooksAreReturned() {
+        when(webClient.method(HttpMethod.GET)).thenReturn(requestBodyUri);
+        when(requestBodyUri.uri(anyString(), anyMap())).thenReturn(requestBody);
+        when(requestBody.retrieve()).thenReturn(response);
+        when(response.bodyToMono(new ParameterizedTypeReference<List<Book>>(){}))
+          .thenReturn(Mono.just(List.of(
+            new Book(1,"Book_1", "Author_1", 1998),
+            new Book(2, "Book_2", "Author_2", 1999)
+          )));
```

Comment:
A couple of things here.

Firstly, it's better to use BDDMockito than Mockito. We're really setting up the givens of this, rather than the whens. In that we're saying "given response's body to mono will return X" ready for the next step of the test to go "when system under test is asked this, then the result is that".

Secondly, I think I mentioned this when we were discussing the outline, you can use deep mocks.

https://www.baeldung.com/mockito-fluent-apis

```java
@Mock(answer = Answers.RETURNS_DEEP_STUBS)
private WebClient webClient;

    @Test
    void givenMockedWebClient_whenAllBooksAreRequested_thenTwoBooksAreReturned() {
        given(webClient.method(HttpMethod.GET)
          .uri(anyString(), anyMap())
          .retrieve()
          .bodyToMono(new ParameterizedTypeReference<List<Book>>(){}))
          .willReturn(Mono.just(List.of(
            new Book(1,"Book_1", "Author_1", 1998),
            new Book(2, "Book_2", "Author_2", 1999)
          )));

```

Reply:
Great stuff @ashleyfrieze, I wasn't aware of this feature, will change it.